DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateGroupsController.cs

```diff
@@ -232,10 +238,155 @@ public IActionResult ConfirmDeleteGroup(int groupId, ConfirmDeleteGroupViewModel
             return RedirectToAction("Index");
         }
 
+        [HttpGet]
+        [Route("Generate")]
+        public IActionResult GenerateGroups()
+        {
+            var registrationFieldOptions = GetRegistrationFieldOptionsSelectList();
+
+            var model = new GenerateGroupsViewModel(registrationFieldOptions);
+            return View(model);
+        }
+
+        [HttpPost]
+        [Route("Generate")]
+        public IActionResult GenerateGroups(GenerateGroupsViewModel model)
+        {
+            if (!ModelState.IsValid)
+            {
+                model.RegistrationFieldOptions = GetRegistrationFieldOptionsSelectList(model.RegistrationFieldOptionId);
+                return View(model);
+            }
+
+            var adminId = User.GetAdminId();
+            var centreId = User.GetCentreId();
+            var isJobGroup = model.RegistrationFieldOptionId == 7;
+            var linkedToField = GetLinkedToFieldValue((int)model.RegistrationFieldOptionId!);
+
+            var (newGroupNames, groupNamePrefix) = GetNewGroupNamesAndPrefix(model, isJobGroup);
+
+            using var transaction = new TransactionScope();
+            foreach (var (newGroupId, newGroupName) in newGroupNames)
+            {
+                var groupName = model.PrefixGroupName ? $"{groupNamePrefix} - {newGroupName}" : newGroupName;
+
+                var groupId = CreateNewGroup(model, adminId, centreId, linkedToField, groupName);
+
+                if (model.AddExistingDelegates && groupId != null)
+                {
+                    groupsService.AddDelegatesWithMatchingAnswersToGroup(
+                        (int)groupId,
+                        linkedToField,
+                        centreId,
+                        isJobGroup ? null : newGroupName,
+                        isJobGroup ? newGroupId : (int?)null
+                    );
+                }
+            }
+
+            transaction.Complete();

```

Comment:
Way to much logic in the Controller that belongs at the service level between this and all the private methods it calls. I would expect this Post action to call a single service method (something like GroupService.GenerateGroupFromRegistrationField) that handles deciding what groups etc.

There is also a load of similar code in GroupService from the other functionality relating to adding delegates to groups, moving delegates between groups etc. that it might be possible to reuse but I'm not 100% sure.

Let me know if you'd like to talk more about what ideally should and shouldn't go in controllers as I know I've commented something similar on another of your tickets

Reply:
Moved all this logic to a service method